packages/fxa-auth-server/fxa-oauth-server/lib/jwt_sub.js

```diff
@@ -0,0 +1,44 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const hex = require('buf').to.hex;
+const hkdf = require('../../lib/crypto/hkdf');
+const config = require('./config');
+
+const PPID_ENABLED = config.get('ppid.enabled');
+const PPID_CLIENT_IDS = new Set(config.get('ppid.enabledClientIds'));
+const PPID_ROTATING_CLIENT_IDS = new Set(config.get('ppid.rotatingClientIds'));
+const PPID_ROTATION_PERIOD_MS = config.get('ppid.rotationPeriodMS');
+const PPID_SALT = config.get('ppid.salt');
+const PPID_INFO = 'oidc ppid sub';
+
+const SUB_LENGTH_BYTES = config.get('unique.id');
+
+module.exports = async function generateSub(
+  userIdBuf,
+  clientIdBuf,
+  clientSeed = 0
+) {
+  const userIdHex = hex(userIdBuf);
+  const clientIdHex = hex(clientIdBuf);
+
+  // TODO - constrain clientSeed.
+
+  if (PPID_ENABLED && PPID_CLIENT_IDS.has(clientIdHex)) {
+    let timeBasedContext = 0;
+    if (PPID_ROTATING_CLIENT_IDS.has(clientIdHex)) {
+      timeBasedContext = Math.floor(Date.now() / PPID_ROTATION_PERIOD_MS);
+    }
+    return hex(
+      await hkdf(
+        `${clientIdHex}.${userIdHex}.${clientSeed}.${timeBasedContext}`,
```

Comment:
So you have two inputs here from the client: one is the seed (`clientSeed`) and the other is the id (`clientIdHex`).  If the intent is that one is a unique ID for the device (or an index into the devices on the account) and the other is given to each device to choose, then this is probably right.

Documenting all of that clearly in the description of the configuration would help.  I'm not following the description you have.

I think that both can be (small) integers, unless you have other constraints on the format of device identifiers that you can rely on.

Reply:
> So you have two inputs here from the client: one is the seed (clientSeed) and the other is the id (clientIdHex). If the intent is that one is a unique ID for the device (or an index into the devices on the account) and the other is given to each device to choose, then this is probably right.

That's right, the idea is that both the addon and FxA can enforce key rotation. The clientId and userId will be the same for a given user all Firefox instances where they have enabled the proxy. The addon could send distinct ppid_seeds for, e.g., private browsing mode and "normal" browsing mode. This would cause distinct subs to be generated for each context. Alternatively, if the secure proxy team wants users to appear distinct on each device, they could send random ppid_seeds on each device and enforce a key change whenever they see fit by changing the seed.